caproto/_log.py

```diff
@@ -108,6 +108,17 @@ def __init__(self, fmt=DEFAULT_FORMAT, datefmt=DEFAULT_DATE_FORMAT,
 
     def format(self, record):
         record.message = record.getMessage()
+        if hasattr(record, 'their_address'):
```

Comment:
This approach works, but it has a side-effect of mutating `record.message`, which could interfere with other formatters (i.e. Elastic). Instead of editing the _message_ we can expand the formatter. I suspect this will also be faster, but the most important feature is that it avoids mutating `record.message`.

There are a couple things to consider here. The "old-style" string formatting notation with `%` is not as flexible as the "new-style" with `'...'.format(...)`. We'll need the new-style in order to format the addresses correctly. That is, the line

```py
 formatted = self._fmt % record.__dict__
```

must be changed to

```py
 formatted = self._fmt.format(**record.__dict__)
```

and the `plain_log_format` and `color_log_format` will need to be updated to use new-style notation, changing from

```py
"%(levelname)1.1s ...
```

to 

```py
'{levelname:1.1s} ...'
```

We can then add the remaining extras fields. The new-style notation will let us format the addresses like `{their_addres[0]}:{their_address[1]}`.

Finally, the question remains how to deal with the fact that some records are missing fields. We could either make a custom formatter that ignores fields that aren't present, or we could customize the `record` to have empty values for any missing tags.

Reply:
Fiddling with this for awhile, the best I have come up with so far is:

```py
In [21]: from collections import ChainMap

In [22]: defaults = {'their_address': ('', ''), 'our_address': ('', ''), 'name': ''}

In [25]: '{role} {name} {their_address[0]}:{their_address[1]} {our_address[0]}:{our_address[1]}'.format_map(ChainMap({'role': 'CLIENT'}, defaults))
Out[25]: 'CLIENT  : :'
```

which still leaves extra spaces and those stray colons. Maybe there's a solution I'm not seeing... if not, we can at least move the `if hasattr(...)` blocks to modify `formatted` rather than modifying `record.message`. That would satisfy our goal to avoid mutating `record.message` and interfering with other formatters. But I think it would be uglier and maybe slower than a nice string-formatting solution if we can find one.

Reply:
The message was mutated. But that doesn't mean we lost key:pair value or field name in Elastic or record.__dict__.
Also I feel mutating message is faster because we avoid search key to see if it's exist when using format_map.

Reply:
I don't think we need to be worried about messing with other formatters as the first line of this formatter is `record.message = record.getMessage()` (which is also the first line of the base class's `format` method https://github.com/python/cpython/blob/9bdd6d1c2a61e0a27dba063e9eccf70e295f73a4/Lib/logging/__init__.py#L664-L677) so I expect that this will be done once per formatter.

Reply:
Good enough for now. It's easy to imagine pushing this to a thread where we have to be concerned about interfering formatters, but we can fix that bug when we hit it.

Reply:
> Also I feel mutating message is faster because we avoid search key to see if it's exist when using format_map.

I'm not sure I get this. Isn't the series of `hasattr` lookups doing the same work?